src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+
+        if ($cursor) {
+            $query
+                ->where($query->expr()->gte('item.sequence', ':cursor'))
+                ->setParameter('cursor', (int) $cursor);
+        }
+
+        $results = $query->execute();
+        $ids = $results->fetchAll(FetchMode::ASSOCIATIVE);
+
+        if (empty($ids)) {
+            return new ItemListPage([], null);
+        }
+
+        if (count($ids) > $limit) {
+            $newCursor = (string) array_pop($ids)['sequence'];
+        }
+
+        $ids = array_map([ItemId::class, 'fromString'], array_column($ids, 'id'));
+
+        return new ItemListPage($ids, $newCursor ?? null);
+    }
+
+    public function count() : int
+    {
+        $results = $this->connection->executeQuery("SELECT COUNT(item.id) count FROM {$this->tableItems} item");
+
+        return (int) $results->fetch()['count'];
+    }
+
+    public function getIterator() : Traversable
+    {
+        while (true) {
+            /** @var Statement $statement */
+            $results = $this->connection->executeQuery(
+                'SELECT version.id, version.version, version.content, version.hash, item.sequence'
+                ." FROM {$this->tableItems} item LEFT JOIN 
+{$this->tableVersions} version ON item.id = version.id AND item.version = version.version AND item.sequence > :sequence"
+                .' ORDER BY item.sequence LIMIT 100',
+                ['sequence' => $cursor ?? 0]
+            );
+
+            while ($result = $results->fetch(FetchMode::ASSOCIATIVE)) {
+                yield $this->rowToItemVersion($result);
+                $cursor = $result['sequence'];
+            }
+
+            if (empty($result)) {
+                break;
+            }
+        }
+    }
+
+    public function getSchema() : Schema
+    {
+        $schema = new Schema();
+
+        $table = $schema->createTable($this->tableItems);
+        $table->addColumn('id', Type::STRING);
+        $table->addColumn('version', Type::INTEGER);
+        $table->addColumn('sequence', Type::INTEGER, ['autoincrement' => true]);
+        $table->setPrimaryKey(['sequence']);
+        $table->addUniqueIndex(['id']);
+
+        $table = $schema->createTable($this->tableVersions);
+        $table->addColumn('id', Type::STRING);
+        $table->addColumn('version', Type::INTEGER);
+        $table->addColumn('content', Type::BLOB);
+        $table->addColumn('hash', Type::STRING);
+        $table->setPrimaryKey(['id', 'version']);
+        $table->addIndex(['id']);
+
+        return $schema;
+    }
+
+    private function rowToItemVersion(array $row) : ItemVersion
+    {
+        ['id' => $id, 'version' => $version, 'content' => $content, 'hash' => $hash] = $row;
+
+        // Some drivers, eg pdo_sqlite, unfortunately return strings rather than resources.
+        // Doctrine doesn't provide an equivalent of `PDOStatement::bindColumn()` either.
+        // This might cause memory problems if the content is too large.
```

Comment:
so what's the type that comes out of this, `@var string|resource`/`@var string|stream`? Worth documenting with a docblock for the variable

Reply:
```php
$platform = $this->connection->getDatabasePlatform();

$id = Type::getType(Type::STRING)->convertToPHPValue($row['id'], $platform);
$version = Type::getType(Type::INTEGER)->convertToPHPValue($row['version'], $platform);
// Some drivers, eg pdo_sqlite, unfortunately return strings rather than resources.
// Doctrine doesn't provide an equivalent of `PDOStatement::bindColumn()` either.
// This might cause memory problems if the content is too large.
$content = Type::getType(Type::BLOB)->convertToPHPValue($row['content'], $platform);
$hash = Type::getType(Type::STRING)->convertToPHPValue($row['hash'], $platform);

return new ItemVersion(ItemId::fromString($id), ItemVersionNumber::fromInt($version), $content, $hash);
```

is technically correct, as integers are also string on some platforms. (These is a little indirection for plain strings...)

Haven't tested what MySQL does.